packages/@react-aria/dialog/docs/useDialog.mdx

```diff
@@ -0,0 +1,211 @@
+to `aria-hidden`, so that only the top-most modal is accessible.
+
+```tsx example
+import {
+  useOverlay,
+  usePreventScroll,
+  useModal,
+  OverlayProvider,
+  OverlayContainer
+} from '@react-aria/overlays';
+import {FocusScope} from '@react-aria/focus';
+import {useButton} from '@react-aria/button';
+
+function ModalDialog({title, children, isOpen, onClose, isDismissable}) {
+  // Handle interacting outside the dialog and pressing
+  // the the Escape key to close the modal. 
+  let ref = React.useRef();
+  let {overlayProps} = useOverlay({
+    ref,
+    onClose,
+    isOpen,
+    isDismissable
+  });
+
+  // Prevent scrolling while the modal is open, and hide content
+  // outside the modal from screen readers.
+  usePreventScroll();
+  useModal();
+
+  // Get props for the dialog and its title
+  let {dialogProps, titleProps} = useDialog({ref});
+
+  return (
+    <div
+      style={{
+        position: 'fixed',
+        zIndex: 100,
+        top: 0,
+        left: 0,
+        bottom: 0,
+        right: 0,
+        background: 'rgba(0, 0, 0, 0.5)',
+        display: 'flex',
+        alignItems: 'center',
+        justifyContent: 'center'
+      }}>
+      <FocusScope contain restoreFocus autoFocus>
+        <div
+          {...overlayProps}
+          {...dialogProps}
+          ref={ref}
+          style={{
+            background: 'white',
+            color: 'black',
+            padding: 30
+          }}>
+          <h3
+            {...titleProps}
+            style={{marginTop: 0}}>
+            {title}
+          </h3>
+          {children}
+        </div>
+      </FocusScope>
+    </div>
+  );
+}
+
+function Example() {
+  let [isOpen, setOpen] = React.useState();
+
+  // useButton ensures that focus management is handled correctly,
+  // across all browsers. Focus is restored to the button once the
+  // dialog closes.
+  let {buttonProps: openButtonProps} = useButton({
+    onPress: () => setOpen(true)
+  });
+
+  let {buttonProps: closeButtonProps} = useButton({
+    onPress: () => setOpen(false)
+  });
+
+  return <>
+    <button {...openButtonProps}>Open Dialog</button>
+    {isOpen &&
+      <OverlayContainer>
+        <ModalDialog 
+          title="Enter your name"
+          isOpen
+          onClose={() => setOpen(false)}
+          isDismissable>
+          <form style={{display: 'flex', flexDirection: 'column'}}>
+            <label>
+              First Name: <input placeholder="John" />
+            </label>
+            <label>
+              Last Name: <input placeholder="Smith" />
+            </label>
+            <button 
+              {...closeButtonProps}
+              style={{marginTop: 10}}>
+              Submit
+            </button>
+          </form>
+        </ModalDialog>
+      </OverlayContainer>
+    }
+  </>;
+}
+
+// Application must be wrapped in an OverlayContainer so that it can be
+// hidden from screen readers when a modal opens.
+<OverlayProvider>
+  <Example />
+</OverlayProvider>
+```
```

Comment:
There is a lot of stuff happening in this example that you cover in the paragraph above but I wonder if it would be better to only expose the `useDialog` parts (though this would probably take some additional work to be able to hide the rest of the code or to rework the example entirely) and pare down the explanation accordingly (or reword it to point out what useDialog is adding here). To compensate for the rest of the explanation, maybe it would be better to link to the `useModal`/`useOverlay`/etc future docs instead. It would also help consolidate all the info about these other hooks into one place so you don't have to re-explain what each hook does (unless those hooks won't each have a dedicated docs page?).

On the other hand this does give a very comprehensive explanation of how to build a dialog from scratch all in one place, something that would be lost if split up, hrm. Lemme know what you think

Reply:
Same for useOverlayTrigger

Reply:
Yeah it is quite complex... Some of the hooks wouldn't be very useful to document on their own though, i.e. the example kinda needs to be a combination of these things. We could create some higher level hooks to combine some of them into a single hook I suppose, but then you lose the option to only include some functionality depending on your use case (or we end up needing to document those parts separately somehow). 🤔

Reply:
I feel like even if we had higher level hooks that combined these items, we would still need to document them anyway. I'd rather be a little over the top than not have a real life example. Plus using aria by itself already means you are a bit of an advanced user.